consensus: Simplify ConnectTrace

jnewbery opened the pull request:
ConnectTrace is now only used to track blocks that were connected during
an ActivateBestChainStep call. Simplify it to a typedef and fix
comments.

Rather than trying to remove ConnectTrace, which would change the order that notifications are fired in the case of a re-org (https://github.com/bitcoin/bitcoin/pull/17562#issuecomment-615251297), just simplify it and fix comments.

jnewbery committed:
[validation] Simplify ConnectTrace

ConnectTrace is now only used to track blocks that were connected during
an ActivateBestChainStep call. Simplify it to a typedef.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
     CBlockIndex* pindex = nullptr;
     std::shared_ptr<const CBlock> pblock;
-    PerBlockConnectTrace() {}
+    PerBlockConnectTrace(CBlockIndex* index, std::shared_ptr<const CBlock> block) : pindex(index), pblock(block) {}
```
```suggestion
    PerBlockConnectTrace(CBlockIndex* index, const std::shared_ptr<const CBlock>& block) : pindex(index), pblock(block) {}
```

style nit to avoid shared_ptr increment/decrement

jnewbery replied:
Thanks for this comment! It sent me down a bit of a rabbit hole about how to pass smart pointers. Here's what I've learned:

I think your suggestion to change the ctor to take a const lvalue reference makes a copy and therefore increases the ref count (the `const` is the callee saying to the caller "I won't steal this reference", and so if the callee wants to keep ownership, it needs to make a copy). We can avoid that copy in three ways:

- pass by non-const lvalue reference (`std::shared_ptr<const CBlock>&`) which is the callee saying "I may or may not steal this reference" and call the constructor with an lvalue (ie `thisBlock` below).
- pass by value (`std::shared_ptr<const CBlock>`), and have the caller cast to rvalue reference in the call (ie `std::move(thisBlock)` below). The `std::move` is the caller saying "I no longer want ownership of this reference. In this case, the caller can still keep shared ownership by passing an lvalue (`thisBlock`), which would increase the ref count.
- pass by rvalue (`std::shared_ptr<const CBlock>&&`) which is the callee saying "I will steal ownership". The caller has to call using an rvalue reference (`std::move(thisBlock)`). Calling by lvalue would be invalid.

I think it's best to keep this as pass-by-value and have the caller pass it with an rvalue reference.

Also, the initializer list needs to use a std::move to prevent the ref increment/decrement, which I'll change.

Some references:
- summary of ways to pass smart pointers as arguments (it's talking about unique_ptr, but the same concepts are true for shared_ptrs): https://stackoverflow.com/questions/8114276/how-do-i-pass-a-unique-ptr-argument-to-a-constructor-or-a-function
- CPP Core Guidelines on passing shared pointers: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-owner
- Herb Sutter on smart pointer parameters: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/
- Scott Meyers on passing move types by value/rvalue reference: http://scottmeyers.blogspot.com/2014/07/should-move-only-types-ever-be-passed.html
- Demonstration code of the different ways to pass a shared_ptr and the effect on ref counts: https://gist.github.com/jnewbery/10ee24a5ce783932accddbe4347f110e

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
-        return blocksConnected;
-    }
-};
+typedef std::vector<PerBlockConnectTrace> ConnectTrace;
```
Why is this needed?

jnewbery replied:
It's not needed. I just wanted to keep it here so the comment made sense and I didn't have to move it to the header file.

MarcoFalke replied:
Didn't see the new comment, but I think you can put the new comment to the single place where connecttrace is instantiated. No strong opinion, though.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2868,7 +2843,7 @@ bool CChainState::ActivateBestChain(BlockValidationState &state, const CChainPar
                 pindexNewTip = m_chain.Tip();
 
-                for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
+                for (auto& trace : connectTrace) {
```
Why is this change to auto? `auto` shouldn't be used in consensus code unless there is a reason

jnewbery replied:
Sounds sensible. I'll revert.

MarcoFalke commented on src/validation.h:
```diff
@@ -409,7 +409,9 @@ enum DisconnectResult
-class ConnectTrace;
+// forward declaration for ActivateBestChainStep and ConnectTip
+struct PerBlockConnectTrace;
+typedef std::vector<PerBlockConnectTrace> ConnectTrace;
```
```suggestion
using ConnectTrace = std::vector<PerBlockConnectTrace>;
```

style nit for C++11

jnewbery replied:
I don't think it makes a difference for aliases which are not templates (and I can't see this in the style guide), but changed anyway.

jonatack replied:
> I don't think it makes a difference for aliases which are not templates (and I can't see this in the style guide), but changed anyway.

This is my understanding as well after looking this up, along with `typedef` being the earlier, traditional syntax of the two. Perhaps one advantage `typedef` may have over `using` -- this is personal observation only and not for bikeshedding purposes :) -- is that it is used only for declaring type aliases.

jonatack replied:
Git grepping instances of type aliases seems easier with`typedef` as well.

maflcko reviewed (commented):
Concept ACK

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
-        // blocks are added after all the conflicted transactions have
-        // been filled in. Thus, the last entry should always be an empty
-        // one waiting for the transactions from the next block. We pop
-        // the last entry here to make sure the list we return is sane.
```
Note to other reviewers: This comment obviously no longer applies after 312d27b

jnewbery commented:
Thanks for the review @MarcoFalke . I've addressed your comments.

jonatack commented:
Concept ACK

robot-dreams commented on src/validationinterface.h:
```diff
@@ -132,14 +132,19 @@ class CValidationInterface {
-     * Provides a vector of transactions evicted from the mempool as a result.
+     *
+     * The mempool is not guaranteed to be in a consistent state when this
+     * notification is fired. Listeners should use UpdatedBlockTip for that.
```
Beginner question: what do you mean by "consistent state" here?

i.e. what could go wrong with the mempool after some, but not all, of the `ActivateBestChainStep` involved in `ActivateBestChain`?

jnewbery replied:
cs_main and mempool.cs are both still being held when the `BlockConnected` signal is called, and `UpdateMempoolForReorg()` may have been called after blocks were connected but before the `BlockConnected` signal is called. See the comment above for the `TransactionAddedToMempool` signal.

The `UpdatedBlockTip` signal is called just before releasing `cs_main`, when we know that the mempool is consistent with the chain tip.

jnewbery replied:
I agree with your first point. For the second point, if we consider the client to be a passive listener (ie the only input it's getting is from the validation interface), then it shouldn't/can't acquire those locks. It should wait for a `UpdatedBlockTip` to be sure that the state is consistent.

robot-dreams reviewed (commented):
Concept ACK, thanks for doing this cleanup!

Unit tests passed locally

(Aside: What's the desired benefit of reviewers doing local testing, given the nice CI system? Is it to get more test runs on different system configurations?)

jnewbery commented on src/validation.cpp:
```diff
@@ -2610,7 +2585,7 @@ bool CChainState::ConnectTip(BlockValidationState& state, const CChainParams& ch
     LogPrint(BCLog::BENCH, "- Connect block: %.2fms [%.2fs (%.2fms/blk)]\n", (nTime6 - nTime1) * MILLI, nTimeTotal * MICRO, nTimeTotal * MILLI / nBlocksTotal);
 
-    connectTrace.BlockConnected(pindexNew, std::move(pthisBlock));
+    connectTrace.emplace_back(pindexNew, pthisBlock);
```
Yes, you're right! The caller should cast to rvalue reference by calling `std::move()` (that was my intention in https://github.com/bitcoin/bitcoin/pull/18685#discussion_r410717136).

Thanks for catching this.

jnewbery commented:
I've addressed the comment here https://github.com/bitcoin/bitcoin/pull/18685#discussion_r479612111 and rebased on master (there wasn't a conflict but this was a few thousand commits behind master and it's a small change).

robot-dreams commented on src/validation.cpp:
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
+ * Used to track blocks that were connected as part of a single
+ * ActivateBestChainStep call. The BlockConnected validationinterface signal is
+ * fired after a single ABCS call for all blocks that were connected in that
+ * ABCS call.
```
Nit (feel free to ignore): How would you feel about rewording this to make it really obvious that there's one BlockConnected signal per block? For example:

    Used to track blocks that were connected as part of a single
    ActivateBestChainStep call. After the ABCS call, a separate BlockConnected
    validationinterface signal is fired for each block that was connected
    in the ABCS call.

jnewbery replied:
Sure. That seems better. Updated.

robot-dreams reviewed (commented):
ACK d956ac72f11c0888c6d719d257aa050a37d12ff2

ajtowns reviewed (commented):
utACK 4faaaa1d5935e1296c624a8d9d414989ec2fede6

jnewbery committed:
[docs] Update comments for BlockConnected and BlockDisconnected

Clarify that the mempool is not guaranteed to be in a consistent state
with the chain tip when BlockConnected and BlockDisconnected are fired.
Remove outdated comment about BlockConnected providing a vector of
transactions.

jnewbery commented on src/validation.cpp:
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
-        return blocksConnected;
-    }
-};
+using ConnectTrace = std::vector<PerBlockConnectTrace>;
```
Done. I've moved everything into the class declaration in the header.

jnewbery commented on src/validation.cpp:
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
     CBlockIndex* pindex = nullptr;
     std::shared_ptr<const CBlock> pblock;
-    PerBlockConnectTrace() {}
+    PerBlockConnectTrace(CBlockIndex* index, std::shared_ptr<const CBlock> block) : pindex(index), pblock(std::move(block)) {}
```
Done.

jnewbery commented:
Thanks for the review, @ajtowns! I've addressed your comments and rebased on master (since this was a couple of thousand commits behind the current master).

jnewbery commented:
@ajtowns do you mind looking at this again? I've addressed your review comments.

jnewbery commented:
Closing due to lack of reviewer interest.

Rspigler commented:
Mark up for grabs?

jnewbery commented:
> Mark up for grabs?

@Rspigler feel free to pick this up if you're interested. I'm happy to review, but this didn't get much reviewer interest last time round.
